Implement URLPattern matching

jelee53 opened the pull request:
#### 719e478475ea019476249826a0c0707bb3f2e72f
<pre>
Implement URLPattern matching
<a href="https://bugs.webkit.org/show_bug.cgi?id=284077">https://bugs.webkit.org/show_bug.cgi?id=284077</a>
<a href="https://rdar.apple.com/140946287">rdar://140946287</a>

Reviewed by Chris Dumez and Sihui Liu.

URLPattern match function spec: <a href="https://urlpattern.spec.whatwg.org/#url-pattern-match">https://urlpattern.spec.whatwg.org/#url-pattern-match</a>

* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.worker-expected.txt:
* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
(WebCore::URLPattern::test const):
(WebCore::URLPattern::exec const):
(WebCore::URLPattern::match const):
* Source/WebCore/Modules/url-pattern/URLPattern.h:
* Source/WebCore/Modules/url-pattern/URLPattern.idl:
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
(WebCore::canonicalizePort):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
(WebCore::URLPatternUtilities::URLPatternComponent::componentExec const):
(WebCore::URLPatternUtilities::URLPatternComponent::createComponentMatchResult const):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.h:
* Source/WebCore/Modules/url-pattern/URLPatternResult.h:

Canonical link: <a href="https://commits.webkit.org/287741@main">https://commits.webkit.org/287741@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d899a99cb98106b33dd6af3cc60c3ca853e6794b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80608 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59615 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34542 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85132 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31590 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82719 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68676 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7921 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62962 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20766 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83677 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53083 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73380 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43267 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50372 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27540 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30049 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71519 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28063 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86565 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7834 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5556 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71255 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8009 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69217 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70495 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14508 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13457 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12501 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7796 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13315 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7635 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11154 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9440 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jelee53 commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -85,5 +92,70 @@ bool URLPatternComponent::matchSpecialSchemeProtocol(ScriptExecutionContext& con
+
+static String valueToWTFString(JSC::JSValue input)
+{
+    // FIXME: Is there a better, compact way to do this?
```
@szewai @cdumez 

Is there a better, shorter way to convert a JSValue to a string?

cdumez replied:
`JSValue::toString()`?

cdumez replied:
`JSValue::toWTFString()` even

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -87,5 +89,40 @@ bool URLPatternComponent::matchSpecialSchemeProtocol(ScriptExecutionContext& con
+
+    auto globalObject = context.globalObject();
+    Ref vm = globalObject->vm();
+    auto length = execResult.get(globalObject, vm->propertyNames->length).toIntegerOrInfinity(globalObject);
```
```
ASSERT(std::isfinite(length));
ASSERT(length >= 0);
```

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -162,8 +163,13 @@ ExceptionOr<String> canonicalizePort(StringView portValue, const std::optional<S
+    if (protocolValue) {
+        auto parsedPort = parseInteger<uint16_t>(portValue);
+        if (!parsedPort || isDefaultPortForProtocol(*parsedPort, *protocolValue))
+            return String { ""_s };
```
If you want to return empty string, it should be `return emptyString()`; if you want to return null string, it should be `return String { }`.

jelee53 replied:
```
URLPatternCanonical.cpp:169:20: error: no viable conversion from returned value of type 'const String' to function return type 'ExceptionOr<String>'
  169 |             return emptyString(); // String { ""_s };
  
  

```

It tried that initially but I get this complaint that emptyString() is const, so it can't be returned.

cdumez replied:
Maybe `return { emptyString() };` or `return ExceptionOr<String> { emptyString() };`.

`String { ""_s };` is really not great.

szewai replied:
Also, remember to update this

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -257,16 +257,24 @@ ExceptionOr<Ref<URLPattern>> URLPattern::create(ScriptExecutionContext& context,
+            return maybeResult.releaseException();
+
+        return maybeResult.releaseReturnValue().has_value();
+    }
```
We normally put error case in the front, like:
```
if (!input)
    return Exception { ExceptionCode::NotSupportedError };

auto maybeResult = match(context, WTFMove(*input), WTFMove(baseURL));
```

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -257,16 +257,24 @@ ExceptionOr<Ref<URLPattern>> URLPattern::create(ScriptExecutionContext& context,
+        if (maybeResult.hasException())
+            return maybeResult.releaseException();
+
+        return maybeResult.releaseReturnValue().has_value();
```
Not need to release, either `return maybeResult.returnValue().has_value();` or `return return !!maybeResult.returnValue();`

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -257,16 +257,24 @@ ExceptionOr<Ref<URLPattern>> URLPattern::create(ScriptExecutionContext& context,
+    if (input)
+        return match(context, WTFMove(*input), WTFMove(baseURL));
 
     return Exception { ExceptionCode::NotSupportedError };
```
Let's check `!input` first

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -324,4 +332,107 @@ ExceptionOr<void> URLPattern::compileAllComponents(ScriptExecutionContext& conte
+
+    bool isURLPatternInit = false;
+    if (std::holds_alternative<URLPatternInput>(input))
+        isURLPatternInit = std::holds_alternative<URLPatternInit>(std::get<URLPatternInput>(input));
```
We might use `WTF::switchOn` to omit some type checks

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -324,4 +332,107 @@ ExceptionOr<void> URLPattern::compileAllComponents(ScriptExecutionContext& conte
+        URL inputAsURL;
+
+        if (std::holds_alternative<URL>(input))
+            inputAsURL = std::get<URL>(input);
```
`if (const URL *result = std::get_if<URL>(&input))`

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -324,4 +332,107 @@ ExceptionOr<void> URLPattern::compileAllComponents(ScriptExecutionContext& conte
+        if (std::holds_alternative<URL>(input))
+            inputAsURL = std::get<URL>(input);
+        else {
+            if (baseURLString.has_value()) {
```
`else if (baseURLString) {`

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -324,4 +332,107 @@ ExceptionOr<void> URLPattern::compileAllComponents(ScriptExecutionContext& conte
+        protocol = inputAsURL.protocol().toString();
+        username = inputAsURL.user();
+        password = inputAsURL.password();
+        hostname = !inputAsURL.host().isEmpty() ? inputAsURL.host().toString() : emptyString();
```
Is `isEmpty()` check necessary? Empty StringView would probably return empty string for `toString()`

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -324,4 +332,107 @@ ExceptionOr<void> URLPattern::compileAllComponents(ScriptExecutionContext& conte
+        username = inputAsURL.user();
+        password = inputAsURL.password();
+        hostname = !inputAsURL.host().isEmpty() ? inputAsURL.host().toString() : emptyString();
+        port = inputAsURL.port().has_value() ? String::number(*inputAsURL.port()) : emptyString();
```
`has_value()` is usually unnecessary when checking if `std::optional` contains a value

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -87,5 +89,40 @@ bool URLPatternComponent::matchSpecialSchemeProtocol(ScriptExecutionContext& con
+        ++index;
+    }
+
+    result.groups = groups;
```
`WTFMove(groups)`;
Or you should just `return URLPatternComponentResult { WTFMove(input), WTFMove(groups) };`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -87,5 +89,40 @@ bool URLPatternComponent::matchSpecialSchemeProtocol(ScriptExecutionContext& con
+    unsigned index = 1;
+    while (index < length) {
+        auto match = execResult.get(globalObject, index);
+        URLPatternComponentResult::PatternMatch matchResult = !match.isNull()
```
`auto`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -87,5 +89,40 @@ bool URLPatternComponent::matchSpecialSchemeProtocol(ScriptExecutionContext& con
+    ASSERT(length >= 0 && std::isfinite(length));
+
+    unsigned index = 1;
+    while (index < length) {
```
Seems like this should be a `for` loop

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -87,5 +89,40 @@ bool URLPatternComponent::matchSpecialSchemeProtocol(ScriptExecutionContext& con
+        ++index;
+    }
+
+    return URLPatternComponentResult { !input.isEmpty() ? WTFMove(input) : ""_s, WTFMove(groups) };
```
`emptyString()` instead of `""_s`

cdumez reviewed (approved):
LGTM with nits.

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.h:
```diff
@@ -31,11 +31,14 @@
 class RegExp;
 class VM;
+class JSValue;
+class JSArray;
```
Is `JSArray` needed?

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -87,5 +89,39 @@ bool URLPatternComponent::matchSpecialSchemeProtocol(ScriptExecutionContext& con
+    return URLPatternComponentResult { !input.isEmpty() ? WTFMove(input) : emptyString(), WTFMove(groups) };
+}
+
+
```
Extra blank line

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -87,5 +89,39 @@ bool URLPatternComponent::matchSpecialSchemeProtocol(ScriptExecutionContext& con
+        URLPatternComponentResult::NameMatchPair pair = match.isNull() || match.isUndefined()
+            ? URLPatternComponentResult::NameMatchPair { m_groupNameList[index - 1], std::monostate { } }
+            : URLPatternComponentResult::NameMatchPair { m_groupNameList[index - 1], match.toWTFString(globalObject) };
+        groups.append(pair);
```
`WTFMove(pair)`?

szewai replied:
May write it as: 
```
std::variant<std::monostate, String> value;
if (match.isNull() || match.isUndefined())
    value = match.toWTFString(globalObject);
URLPatternComponentResult::NameMatchPair pair { m_groupNameList[index - 1], WTFMove(value) };
```

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -87,5 +89,39 @@ bool URLPatternComponent::matchSpecialSchemeProtocol(ScriptExecutionContext& con
+// https://urlpattern.spec.whatwg.org/#create-a-component-match-result
+URLPatternComponentResult URLPatternComponent::createComponentMatchResult(ScriptExecutionContext& context, String&& input, const JSC::JSValue& execResult) const
+{
+    URLPatternComponentResult result;
```
Is this used?

szewai reviewed (approved):
r=me with nits

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -324,4 +331,127 @@ ExceptionOr<void> URLPattern::compileAllComponents(ScriptExecutionContext& conte
+    protocol = input.protocol().toString();
+    username = input.user();
+    password = input.password();
+    hostname = !input.host().isNull() ? input.host().toString() : emptyString();
```
Can we just do `hostname = input.host().toString()`? Do we really distinguish between empty string and null string in the existing implementation?

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -324,4 +331,127 @@ ExceptionOr<void> URLPattern::compileAllComponents(ScriptExecutionContext& conte
+        ASSERT(!inputURL->isEmpty() && inputURL->isValid());
+        matchHelperAssignInputsFromURL(*inputURL, protocol, username, password, hostname, port, pathname, search, hash);
+        String urlAsString = inputURL->string();
+        result.inputs = Vector<URLPatternInput> { WTFMove(urlAsString) };
```
`result.inputs = Vector<URLPatternInput> { inputURL->string() };`?

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -324,4 +331,127 @@ ExceptionOr<void> URLPattern::compileAllComponents(ScriptExecutionContext& conte
+        String urlAsString = inputURL->string();
+        result.inputs = Vector<URLPatternInput> { WTFMove(urlAsString) };
+    } else {
+        enum InputSwitchReturnTypes { ReturnTypeVoid, ReturnTypeNull, ReturnTypeException };
```
`InputSwitchReturnTypes` => `InputSwitchReturnType`, and you may not need this.
Just have a `ExceptionOr<bool> hasError = false;` and update it as needed in the `switchOn` statements, like:
```
If (hasError.hasException())
    return inputCheckResult.releaseException();
if (hasError.returnValue())
    return { std::nullopt };
```

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -324,4 +331,127 @@ ExceptionOr<void> URLPattern::compileAllComponents(ScriptExecutionContext& conte
+        URLPatternInput* inputPattern = std::get_if<URLPatternInput>(&input);
+
+        auto switchOnReturn = WTF::switchOn(*inputPattern,
+            [&] (URLPatternInit value) -> InputSwitchReturnTypes {
```
`const URLPatternInit& value`

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -324,4 +331,127 @@ ExceptionOr<void> URLPattern::compileAllComponents(ScriptExecutionContext& conte
+    }
+
+    auto protocolExecResult = m_protocolComponent.componentExec(context, protocol);
+    if (protocolExecResult.isNull())
```
Do we also need to check if the result is `undefined`?

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -324,4 +331,127 @@ ExceptionOr<void> URLPattern::compileAllComponents(ScriptExecutionContext& conte
+                matchHelperAssignInputsFromInit(processedInit, protocol, username, password, hostname, port, pathname, search, hash);
+
+                return InputSwitchReturnTypes::ReturnTypeVoid;
+            }, [&] (String value) -> InputSwitchReturnTypes {
```
const String& value

jelee53 committed:
Implement URLPattern matching
https://bugs.webkit.org/show_bug.cgi?id=284077
rdar://140946287

Reviewed by Chris Dumez and Sihui Liu.

URLPattern match function spec: https://urlpattern.spec.whatwg.org/#url-pattern-match

* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.worker-expected.txt:
* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
(WebCore::URLPattern::test const):
(WebCore::URLPattern::exec const):
(WebCore::URLPattern::match const):
* Source/WebCore/Modules/url-pattern/URLPattern.h:
* Source/WebCore/Modules/url-pattern/URLPattern.idl:
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
(WebCore::canonicalizePort):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
(WebCore::URLPatternUtilities::URLPatternComponent::componentExec const):
(WebCore::URLPatternUtilities::URLPatternComponent::createComponentMatchResult const):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.h:
* Source/WebCore/Modules/url-pattern/URLPatternResult.h:

Canonical link: https://commits.webkit.org/287741@main
